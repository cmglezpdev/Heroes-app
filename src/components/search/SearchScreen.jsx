import { useNavigate, useLocation } from "react-router-dom";
import queryString from 'query-string'

import { useForm } from "../../Hooks/useForm";
import getHeroByName from "../../selectors/getHeroNyName";
import { HeroCard } from "../hero/HeroCard";
import { useMemo } from "react";


export const SearchScreen = () => {

  const navigate = useNavigate();
  const location = useLocation();
  const { q = '' } = queryString.parse(location.search);

  const [values, handleInputChange] = useForm({
    searchText: q
  });

  const { searchText } = values;
  const heroesFiltered = useMemo( () => getHeroByName(q), [q]);


    const handleSubmit = ( event ) => {
      event.preventDefault();
      navigate(`?q=${searchText}`);
  
      
    }

  

    return (
      <>
        <h1>Search</h1>
        <hr />

        <div className="row">
          <div className="col-5">
            <h4>Buscar</h4>
            <hr />

            <form onSubmit={ handleSubmit }>
              <input 
                type="text" 
                name="searchText" 
                className="form-control"
                placeholder="Buscar un heroe"
                autoComplete="off"
                value={ searchText }
                onChange={handleInputChange}
              />

              <button
                type="submit"
                className="btn btn-outline-primary mt-1"
              >
                Buscar...
              </button>

            </form>
          </div>

          <div className="col-7">
            <h4>Resutados</h4>
            <hr />

            {
              (q === '') 
              ? <div className="alert alert-info">Buscar un heroe</div>
              : (heroesFiltered.length == 0)
                  && <div className="alert alert-danger">No hay resultados de: { q }</div>
            }

            {
              heroesFiltered.map(hero => (
                <HeroCard
                  key={hero.id}
                  { ...hero } 
                />
              ))
            }
          </div>

        </div>
      </>
    )
}
