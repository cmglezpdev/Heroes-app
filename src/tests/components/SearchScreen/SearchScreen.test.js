/* eslint-disable testing-library/no-node-access */
/* eslint-disable testing-library/prefer-screen-queries */
import { fireEvent, render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';
import getHeroByName from '../../../selectors/getHeroNyName'

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}))

describe('Tests abput SearchScreen', () => { 

    test('should show it correctly with default values', () => {

        const { container } = render(
                <MemoryRouter initialEntries={['/search']}>
                    <SearchScreen />
                </MemoryRouter>
        );

        expect( container ).toMatchSnapshot();
        screen.getByText('Buscar un heroe');
    });

    test('should show to Batman and the input with queryString Value', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        expect(input).toHaveProperty('value', 'batman');
    })

    test('should show a error if not exist a hero with the name', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=123432']}>
                <SearchScreen />
            </MemoryRouter>
        );
        
        screen.getByText('No hay resultados de: 123432');
    })

    test('should show the same images that heroes in the search result', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=man']}>
                <SearchScreen />
            </MemoryRouter>
        );
      
        const heroes = getHeroByName('man');
        const imgs = screen.getAllByRole('img');
        
        expect(heroes.length).toBe(imgs.length);
        
        const comp1 = heroes.map(hero => hero.superhero);
        const comp2 = imgs.map(img => img.getAttribute('alt'));
        comp1.sort();
        comp2.sort();

        expect(comp1).toEqual(comp2);
    });

    test('should called the navigate for to change the url', () => {
        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        );
    
        const input = screen.getByRole('textbox');
        fireEvent.change(input, { 
            target: {
                name: 'searchText', 
                value: 'superman' 
            } 
        })
        
        const form = screen.getByRole('textbox').parentElement;
        fireEvent.submit(form);
        expect(mockNavigate).toHaveBeenCalledTimes(1);
        expect(mockNavigate).toHaveBeenCalledWith('?q=superman');
    });

 })