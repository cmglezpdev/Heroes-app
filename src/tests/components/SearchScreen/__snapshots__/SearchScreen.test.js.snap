// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests abput SearchScreen should show it correctly with default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form>
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Buscar un heroe"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
          type="submit"
        >
          Buscar...
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Resutados
      </h4>
      <hr />
      <div
        class="alert alert-info"
      >
        Buscar un heroe
      </div>
    </div>
  </div>
</div>
`;
